### PR TITLE
qemu-guest-agent: Add command 'guest-get-disk' for checking disks info

### PR DESCRIPTION
**Pull request by 6-dehan**
Add command 'guest-get-disk' for checking disk_name,
disk_dependencies, disk_partition, disk_alias(lvm only).

ID: 1906654
Signed-off-by: Dehan Meng <demeng@redhat.com>

### CODE REVIEW

**Comment by 6-dehan** [comment]
@vivianQizhu please help to review this new command.
(1/1) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_get_disks.q35: PASS (55.59 s)

File: qemu/tests/qemu_guest_agent.py [hunk 1]
```diff
@@ -1224,6 +1224,81 @@ def get_allocation_bitmap():
...
+        for disk_info in disks_info_qga:
+            diskname = disk_info['name']
+            cmd_diskinfo_guest = params['diskinfo_guest_cmd'] % diskname
+            if disk_info['name'] in raw_disk_guest:
+                cmd_rawdisk_info_guest = params['rawdisk_info_guest_cmd'] % diskname
+                pre_disk_info_guest = session.cmd_output(cmd_rawdisk_info_guest)
+                disk_info_guest = _process_output(pre_disk_info_guest)
+            else:
+                pre_disk_info_guest = session.cmd_output(cmd_diskinfo_guest)
+                disk_info_guest = _process_output(pre_disk_info_guest)
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:1265]:
```suggestion
                cmd_rawdisk_info_guest = params['rawdisk_info_guest_cmd'] % diskname
                disk_info_guest = session.cmd_output(cmd_rawdisk_info_guest)
            else:
                disk_info_guest = session.cmd_output(cmd_diskinfo_guest)
            disk_info_guest = eval(disk_info_guest.strip().strip(','))
```

File: qemu/tests/qemu_guest_agent.py [hunk 2]
```diff
@@ -1224,6 +1224,81 @@ def get_allocation_bitmap():
...
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environment.
+        """
+        def _process_output(pre_output):
+            if pre_output.strip()[-1] == ",":
+                output = eval(pre_output.strip()[:-1])
+                return output
+            else:
+                output = eval(pre_output.strip())
+                return output
```
[the remark this answers is not in the document]
Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:1246]: oh,yes, stupid issue I made. thanks

**Comment by 6-dehan** [comment]
@vivianQizhu @kimi1978wy 
all recommendations are updated. please have a check.
(3/3) Host_RHEL.m8.u4.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_get_disks.q35: PASS (107.98 s)

File: qemu/tests/qemu_guest_agent.py [hunk 3]
```diff
@@ -1224,6 +1224,73 @@ def get_allocation_bitmap():
...
+            else:
+                if disk_info['dependencies'][0] != disk_info_guest['pkname']:
+                    test.fail("Disk %s dependencies is different "
+                              "between guest and qga." % diskname)
+
+            error_context.context("Check partition", logging.info)
+            if disk_info_guest['type'] != "part":
+                partition = False
+            else:
+                partition = True
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:1281]:
```suggestion
            partition = False if disk_info_guest['type'] != "part" else True
```
But I am okay if you prefer to keep it.

File: qemu/tests/qemu_guest_agent.py [hunk 4]
```diff
@@ -1224,6 +1224,73 @@ def get_allocation_bitmap():
...
+
+        error_context.context("Check all disks info in a loop.",
+                              logging.info)
+        disks_info_qga = self.gagent.get_disks()
+        cmd_get_rawdisk_guest = params['get_rawdisk_guest_cmd']
+        raw_disk_guest = session.cmd_output(cmd_get_rawdisk_guest)
+        for disk_info in disks_info_qga:
+            diskname = disk_info['name']
+            if diskname in raw_disk_guest:
+                cmd_rawdisk_info_guest = params['rawdisk_info_guest_cmd'] % diskname
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:1252]: This line is too long.

File: qemu/tests/qemu_guest_agent.py [hunk 5]
```diff
@@ -1224,6 +1224,73 @@ def get_allocation_bitmap():
...
+            if disk_info['name'] != disk_info_guest['kname']:
+                test.fail("Disk %s name is different "
+                          "between guest and qga." % diskname)
+
+            error_context.context("Check dependencies", logging.info)
+            if disk_info_guest['type'] == 'disk':
+                dependencies = 'null'
+                try:
+                    if disk_info['dependencies'][0] != dependencies:
+                        test.error("Disk %s dependencies should be null." % diskname)
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:1269]: So does this line.
You might use some tools to help you restrict the line length, like add indicator in your editor or IDE.

File: qemu/tests/qemu_guest_agent.py (hunk 3 above)
Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:1281]: you're suggestion make the code looks better and clearer. of course I prefer to use this.

File: qemu/tests/qemu_guest_agent.py (hunk 5 above)
Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:1269]: okay, thanks.

**Comment by 6-dehan** [comment]
@vivianQizhu new code has been pushed.

**Comment by vivianQizhu** [comment]
@kimi1978wy Do you have any more comments?

**Comment by kimi1978wy** [comment]
@vivianQizhu LGTM now, no new comments. @6-dehan pls update the latest test results, thanks

**Comment by 6-dehan** [comment]
@vivianQizhu @kimi1978wy 
(1/1) Host_RHEL.m8.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_get_disks: PASS (70.99 s)

**Comment by kimi1978wy** [comment]
ACK

Yu Wang <wyu@redhat.com>

**Comment by vivianQizhu** [comment]
Thanks @6-dehan @kimi1978wy

**Comment by 6-dehan** [comment]
> Thanks @6-dehan @kimi1978wy

thanks for your effort and time. @vivianQizhu @kimi1978wy

**Pull request merged**